[DOCS-9915] Mobile App Performance updates for iOS and Android (#27404)
DataDog/documentation: content/en/real_user_monitoring/mobile_and_tv_monitoring/ios/monitoring_app_performance.md

---
title: Android Monitoring App Performance
further_reading:

> - link: https://github.com/DataDog/dd-sdk-android

Reviewer: ```suggestion
- link: https://github.com/DataDog/dd-sdk-ios
```

  tag: "Source Code"

>   text: Source code for dd-sdk-android

Reviewer: ```suggestion
  text: Source code for dd-sdk-ios
```

- link: /real_user_monitoring
  tag: Documentation
  text: Explore Datadog RUM
---
## Overview

> View timings help you understand the performance of your application from a user's perspective. RUM comes with both out-of-the-box automatic timings (Time-to-Network-Settled and Interaction-to-Next-View), as well as a precise API for notifying that the view has finished loading (as only you, as a developer, can know this with certainty).

Author: ```suggestion
View timings help you understand the performance of your application from a user's perspective. RUM comes with both out-of-the-box automatic timings (`Time-to-Network-Settled` and `Interaction-to-Next-View`), as well as a precise API for notifying that the view has finished loading (as only you, as a developer, can know this with certainty).
```

### Time to network settled

The **Time-to-Network-Settled (TNS)** measures the time it takes for a view to be fully loaded with all relevant network calls initiated at the start of the view. TNS is represented by the `@view.network_settled_time` attribute in RUM view events.

By default, TNS is calculated as the time elapsed between the start of the view and the completion of all resources that started within 100ms of the view's start. This behavior is controlled by the `TimeBasedTNSResourcePredicate`, which classifies such resources as "initial."

> To customize the default 100 ms threshold for TNS calculation, you can adjust the threshold value in the `TimeBasedTNSResourcePredicate` and set it for the `networkSettledResourcePredicate` configuration option. This allows you to include resources that start within a custom time window after the view starts:

Author: ```suggestion
To customize the default 100ms threshold for TNS calculation, you can adjust the threshold value in the `TimeBasedTNSResourcePredicate` and set it for the `networkSettledResourcePredicate` configuration option. This allows you to include resources that start within a custom time window after the view starts:
```

### Add your own performance timing

In addition to RUM's default attributes, you can measure where your application is spending its time by using the `addTiming(name:)` API. The timing measure is relative to the start of the current RUM view.

For example, you can time how long it takes for your hero image to appear:

{{< tabs >}}
{{% tab "Swift" %}}
```swift
func onHeroImageLoaded() {
    let rum = RUMMonitor.shared()
    rum.addTiming(name: "hero_image")
}
```
{{% /tab %}}
{{% tab "Objective-C" %}}
```objective-c
- (void)onHeroImageLoad {
    [[DDRUMMonitor shared] addTimingWithName:@"hero_image"];
}
```
{{% /tab %}}
{{< /tabs >}}

Once you set the timing, it is accessible as `@view.custom_timings.<timing_name>`. For example, `@view.custom_timings.hero_image`.

To create visualizations in your dashboards, [create a measure][1] first.

## Understanding performance timings
All view timings are measured relative to the view's start. The exact moment a view starts depends on the type of instrumentation used for tracking views. For more details, see [Views instrumentation versus app lifecycle][2].

## Troubleshooting

> When using the default TimeBasedInitialResourceIdentifier and TimeBasedInteractionIdentifier, TNS and INV timings may be missing in specific cases:

Author: ```suggestion
When using the default `TimeBasedInitialResourceIdentifier` and `TimeBasedInteractionIdentifier`, TNS and INV timings may be missing in specific cases:
```
